src/mpid/ch4/src/ch4i_workq.h

```diff
@@ -0,0 +1,303 @@
+                                                  const void *send_buf,
+                                                  void *recv_buf,
+                                                  MPI_Aint count,
+                                                  MPI_Datatype datatype,
+                                                  int rank,
+                                                  int tag,
+                                                  MPIR_Comm *comm_ptr,
+                                                  int context_offset,
+                                                  int ep_idx,
+                                                  MPI_Status *status,
+                                                  MPIR_Request *request)
+{
+    MPIDI_workq_elemt_t* pt2pt_elemt = NULL;
+    pt2pt_elemt = MPL_malloc(sizeof (*pt2pt_elemt));
+    pt2pt_elemt->op       = op;
+    pt2pt_elemt->send_buf = send_buf;
+    pt2pt_elemt->recv_buf = recv_buf;
+    pt2pt_elemt->count    = count;
+    pt2pt_elemt->datatype = datatype;
+    pt2pt_elemt->rank     = rank;
+    pt2pt_elemt->tag      = tag;
+    pt2pt_elemt->comm_ptr = comm_ptr;
+    pt2pt_elemt->context_offset = context_offset;
+    pt2pt_elemt->status  = status;
+    pt2pt_elemt->request  = request;
+    MPIDI_workq_enqueue(&comm_ptr->dev.work_queues[ep_idx].pend_ops, pt2pt_elemt);
+}
+
+static inline void MPIDI_workq_rma_enqueue_body(MPIDI_workq_op_t op,
+                                                const void *origin_addr,
+                                                int origin_count,
+                                                MPI_Datatype origin_datatype,
+                                                int target_rank,
+                                                MPI_Aint target_disp,
+                                                int target_count,
+                                                MPI_Datatype target_datatype,
+                                                MPIR_Win *win_ptr,
+                                                int ep_idx)
+{
+    MPIDI_workq_elemt_t* rma_elemt = NULL;
+    rma_elemt = MPL_malloc(sizeof (*rma_elemt));
+    rma_elemt->op               = op;
+    rma_elemt->origin_addr      = origin_addr;
+    rma_elemt->origin_count     = origin_count;
+    rma_elemt->origin_datatype  = origin_datatype;
+    rma_elemt->target_rank      = target_rank;
+    rma_elemt->target_disp      = target_disp;
+    rma_elemt->target_count     = target_count;
+    rma_elemt->target_datatype  = target_datatype;
+    rma_elemt->win_ptr          = win_ptr;
+    MPIDI_workq_enqueue(&win_ptr->dev.work_queues[ep_idx].pend_ops, rma_elemt);
+}
+
+static inline int MPIDI_workq_ep_progress_body(int ep_idx)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_workq_elemt_t* workq_elemt = NULL;
+    MPIDI_workq_list_t *cur_workq;
+    MPIDI_WORKQ_PROGRESS_START;
+    MPL_DL_FOREACH(MPIDI_CH4_Global.ep_queues[ep_idx], cur_workq) {
```

Comment:
@halimamer I'm writing this again,and you may be already aware of it, but I believe `ep_queues` needs to be protected by a lock or something.